src/backend/columnar/columnar_tableam.c

```diff
@@ -1589,13 +1592,46 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 								errmsg("only btree and hash indexes are supported on "
 									   "columnar tables ")));
 			}
+
+			/*
+			 * Since we don't support parallel scans on columnar tables, we
+			 * should dissuade postgres from choosing parallel scan for index
+			 * builds. Otherwise, CONCURRENT index builds would leave
+			 * an invalid index behind. Note that we want to do the same for
+			 * non-concurrrent index builds too. This is because, we want to
+			 * fall-back to sync scan instead of throwing an error.
+			 */
+			oldMaxParallelMaintenanceWorkers = max_parallel_maintenance_workers;
+			SetLocalMaxParallelMaintenanceWorkers(0);
 		}
 
 		RelationClose(rel);
 	}
 
 	PrevProcessUtilityHook(pstmt, queryString, context,
 						   params, queryEnv, dest, completionTag);
+
+	if (oldMaxParallelMaintenanceWorkers != -1)
+	{
+		/* altered by current command, roll it back */
+		SetLocalMaxParallelMaintenanceWorkers(oldMaxParallelMaintenanceWorkers);
```

Comment:
Postgres follow this pattern to temporarily override GUCs (e.g. in extensions, functions):
```c
int save_nestlevel = NewGUCNestLevel();
(void) set_config_option("max_parallel_maintenance_workers", settingStr->data,
                         PGC_USERSET, PGC_S_SESSION,
                         GUC_ACTION_SAVE, true, 0, false);
...
AtEOXact_GUC(true, save_nestlevel);
```

Reply:
Thanks, this is much more cleaner and seems to be working.

On the other hand, I'm still not sure how can we enforce not using parallel workers for concurrent index builds too, since postgres commits & starts a new transaction a few times for "concurrent" builds.
This was also a problem with  `SetLocalMaxParallelMaintenanceWorkers` but wanted to note here too.

One possible solution might be manually doing `max_parallel_maintenance_workers = 0` within a try/catch block but sounds like a bad idea.

Reply:
One other option could be injecting `parallel_workers=0` to `CREATE TABLE .. USING columnar` commands.

If we ever decide supporting parallel scan for columnar tables, then we can just run `ALTER TABLE .. RESET ( parallel_workers );` during uprades or so.

Reply:
The try/catch around a large amount of code isn't safe. We could do some kind of transaction hook, but that seems ugly as well.

For the case of `CONCURRENTLY`, I think we should just save the global variables for the GUC, change them, and change them back after the command. An error can leave parallelism unexpectedly disabled for that session. Unless we have a cleaner solution, we can just file an issue until we support parallelism.

Reply:
> An error can leave parallelism unexpectedly disabled for that session.

umm, are we okay with that side effect ? :/

Reply:
Well, the official answer is "no", but I also don't want to expend a lot of design effort or code complexity solving this problem right now. We've gone a couple days without a very appealing solution here (unless I missed something?). One way or another, we need to fix it, and it will be fixed when we support parallelism.

I appreciate you taking these issues seriously though. Please maintain your high standards; I just don't want to get too bogged down on this one issue.

We need a comment clarifying that the problem can happen when a `CREATE INDEX ... CONCURRENTLY` encounters an error, and also note that we should remove the hack when we support parallelism. For regular index builds, we should disable parallelism using the pattern Marco mentioned above.

Another potential shortcut might be to support parallelism in some kind of degenerate way, where one worker still does everything. That would also fix the problem.

Reply:
>Another potential shortcut might be to support parallelism in some kind of degenerate way, where one worker still does everything. That would also fix the problem.

I will give that a try on Monday. If I cannot make it work, then I will fall-back to the idea of directly setting `max_parallel_maintenance_workers`.

Reply:
I think I made that work: 84f9c56e2732aa57d01c66de78b91401f474a381.
Now we support parallelism in a degenerated way but also skip background workers when building the index.